### PR TITLE
Auto tests

### LINKED ISSUE #34: E2E testing tools for Soroban contracts
### Context

During the development of the first smart contracts, we manually tested them in [Futurenet](https://soroban.stellar.org/docs/getting-started/deploy-to-futurenet)

* [Initial raffle contract](https://github.com/eigerco/nebula/issues/5#issuecomment-1644065962)
* [Initial voting contract](https://github.com/eigerco/nebula/issues/5#issuecomment-1645208546)

As we can observe in the above manual tests, the [CLI code](https://github.com/stellar/soroban-tools/blob/main/cmd/soroban-cli/src/commands/contract/deploy.rs) is able to interact with `Futurnet`. That could provide some inspiration.

Futurenet is currently the only environment that runs the Soroban smart contract platform. Its in continuous change. Thats also an extra motivation for having e2e tests.

### Goals

To be able to write a fairly efficient e2e tests suite that works in `Futurenet` , capable of running in a CI.

* **First** Investigate the state of the art of e2e testing tools for Soroban ecosystem.
* Progressively create the tools while we develop the smart contracts, by refining abstractions and use cases.
* To keep an eye in the development process regarding this research idea and log on this issue all the thoughts, abstractions, experiments, lets use this use as communication hub.

### PR DESCRIPTION
**Pull request by mariopil**
**Referenced issue**
Closes #34 

**Summary of changes**
Here's a simple node.js tool for automated contract tests. What it does:
- creates accounts
- funds them using friendbot
- deploys a contract
- runs its methods according to prepared scenario
- checks results and calculates fees

To run it call `npm run tests`. You can specify which tests to run, e.g. `npm run tests voting`. If any of the test fails, tool will exit with -1 code. Tool could be added to GitHub jobs as an action run after each commit. 

Each contract has its separate ts code with the test scenario. To add new test for a new contract just create new ts file, extend the `BaseTests` class and update `TestsRunner` class.

The `Raffle` test is not completed yet, as the contract fails on `futurenet`.

**Reviewer recommendations**

How to make the review easier. i.e "Its a big PR and the recommendation is to go through each commit" 
or "In order to run the tests, execute ..."

### CODE REVIEW

File: tests/.vscode/launch.json
Comment (geofmureithi) [inline tests/.vscode/launch.json:1]: Is this file needed?

Comment (mariopil) [reply tests/.vscode/launch.json:1]: No, I'll remove it.

File: tests/contracts/basetests.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+          const url = `${this.config.friendbot}?addr=${id}`
+          const res = await fetch(url)
+          const json = await res.json()
+
+          if (this.config.showLogs) {
+            console.log(json)
+          }
+        },
+        error => {
+          resultOk = false
```
Comment (geofmureithi) [inline tests/contracts/basetests.ts:55]: This continues executing even if the result is false. Why not use [Promises.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)

File: tests/contracts/basetests.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      '--wasm',
+      `../target/wasm32-unknown-unknown/release/${contract}.wasm`,
+      '--network',
+      this.config.network,
+    ]).then(
+      data => {
+        this.contractId = data.toString().trim()
+      },
+      error => {
+        result = false
```
Comment (geofmureithi) [inline tests/contracts/basetests.ts:80]: Just return the promise and handle the result upstream

File: tests/contracts/raffletests.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+export class RaffleTests extends BaseTests {
+  private tokenId: string = ''
+
+  constructor(protected config: AppConfig) {
+    super(config)
+  }
+
+  async getTokenId() {
+    let resultOk = true
```
Comment (geofmureithi) [inline tests/contracts/raffletests.ts:12]: Same here, return the promise itself instead of bool

File: tests/contracts/raffletests.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        this.logs.push(error.toString())
+        resultOk = false
+        console.error('error: ', error.toString())
+      }
+    )
+    return resultOk
+  }
+
+  async invokeBuyTickets() {
+    let resultOk = true
```
Comment (geofmureithi) [inline tests/contracts/raffletests.ts:71]: Same here with `Promises.all`

File: tests/contracts/raffletests.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            console.error('error: ', error.toString())
+          }
+        )
+      }
+    }
+    return resultOk
+  }
+
+  async invokePlayRaffle() {
+    let resultOk = true
```
Comment (geofmureithi) [inline tests/contracts/raffletests.ts:105]: Same here

File: tests/contracts/votingtests.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+      })
+    return resultOk
+  }
+
+  async calculateFees() {
+    await this.calculateFeesForAccount('admin')
+    await this.calculateFeesForAccount('player_1')
+  }
+
+  public async run(): Promise<boolean> {
```
Comment (geofmureithi) [inline tests/contracts/votingtests.ts:142]: If these methods return promises then you can do:
```js
this.createAccounts()
   .then(res => this.deployContract('voting'))
   .then(......)
   .catch((e) => ... )
```

File: tests/testsrunner.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+  public async runTests(): Promise<boolean> {
+    const specificTests = this.getSpecificTests()
+    let testsToRun = specificTests.length
+    let succeededTests = 0
+    for (const testName of specificTests) {
+      const test = this.tests.get(testName)
+      if (test === undefined) {
+        console.error(`Could not find test ${testName}`)
+        --testsToRun
```
Comment (geofmureithi) [inline tests/testsrunner.ts:24]:
```suggestion
        throw new Error(`Could not find test ${testName}`)
```

**Review by geofmureithi — CHANGES_REQUESTED** [review]
I have left some thoughts on the API.
Let me know what you think.

File: tests/contracts/basetests.ts [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+      )
+    }
+    return resultOk
+  }
+
+  async deployContract(contract: string) {
+    let result = true
+    await this.runSoroban([
+      'contract',
+      'deploy',
```
Comment (geofmureithi) [inline tests/contracts/basetests.ts:68]: I am curious if we can use `install` instead of `deploy` which is the defacto for tests.

File: tests/testsrunner.ts (hunk 7 above)
Comment (mariopil) [reply tests/testsrunner.ts:24]: That would cause the app to exit which I don't want. Let it still continue with other found tests.

File: tests/contracts/basetests.ts (hunk 8 above)
Comment (mariopil) [reply tests/contracts/basetests.ts:68]: I'm not sure what's the difference here, but the idea is to test the contracts as like they were deployed on mainnet, including fee calculations.

File: tests/testsrunner.ts (hunk 7 above)
Comment (geofmureithi) [reply tests/testsrunner.ts:24]: I think the problem inherently comes from having a custom test runner.
Have you considered something like [TestRunner](https://nodejs.org/api/test.html)?
It would help run tests per file, this gives the person running tests ability to control what someone runs.
WDYT?

Comment (mariopil) [reply tests/testsrunner.ts:24]: I have not. Will take a look at it later. The implemented test runner is very simple and I'm not sure there's any need to change it or make it more complex :)

Comment (geofmureithi) [reply tests/testsrunner.ts:24]: I think you are reinventing the wheel. 
Eg I can't run a single test with the current implementation, I have to run all or change the code. 
Test runner comes bundled with node and is well documented and tested. I don't think that is complexity as compared to having to reference a custom impl. The changes requested are not major, you just have to wrap the calls eg:
```js
test('Raffle Contract', async (t) => {
  await t.test('init' , (t) => {
    assert.strictEqual(1, 1);
  });

  await t.test('create_proposal;', (t) => {
    assert.strictEqual(2, 2);
  });
}); 
``` 
Now:
- I can run each file alone.
- The test runner manages what fails etc.
- The test runner can output other file formats if needed.

If you are not into working on it, I can do it because I don't feel ok with the current impl.

Comment (mariopil) [reply tests/testsrunner.ts:24]: I think you misunderstand the idea of this tool. Its purpose is not to run each of the method alone - for this you have the rust unit tests. Idea of this tool is to test the specific scenario. Currently implemented scenarios are based on the tests Eloy did manually to test the contracts (https://github.com/eigerco/nebula/issues/5). 
And yes if you want to test different scenario - you have to implement another file. Treat the single test file as an integration test case. Actually it might be a good idea to change the test file names as currently they don't tell what is actually being tested.

Comment (geofmureithi) [reply tests/testsrunner.ts:24]: > Its purpose is not to run each of the method alone

I am not suggesting any method to be run alone:
```js
public async run(): Promise<boolean> {
    let result = await this.createAccounts()
    if (!result) return false
    result = await this.deployContract('raffle')
    // ....
}
  ```
  becomes
  ``` js
test('Raffle Contract', async (t) => {
    const contract = new Contract();
    await t.test('createAccounts' , async (t) => {
        await contract.createAccounts()
    });

    await t.test('deployContract', async (t) => {
      await contract.deployContract('raffle')
    });
    .....
});
```
Everything is run the same, but we get a ready test API that is documented and extensible, and we don't have to manage it

Comment (mariopil) [reply tests/testsrunner.ts:24]: There's nothing more to manage than in the example you've posted. You still have to write at least the same amount of code. I don't see a value added here.

Comment (geofmureithi) [reply tests/testsrunner.ts:24]: I think I disagree with the notion that it adds no value.
First of all we wouldn't have the `testrunner` file, secondly we don't need to declare a new contract in code, just add the file and it runs.
We plan to have up to 10 contracts, I think I see a lot of value in this.
I can do a PR with the changes I am recommending if need be.

Comment (geofmureithi) [reply tests/testsrunner.ts:24]: Please see https://github.com/eigerco/nebula/pull/47

**Comment by geofmureithi** [comment]
Are these tests run by CI?

File: tests/appconfig.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import configJson from './config.json'
+
+export class AppConfig {
+  public friendbot: string = ''
```
Comment (geofmureithi) [inline tests/appconfig.ts:4]:
```suggestion
  public friendbot: string = configJson.server.friendbot;
```
Is this possible? We would not need the `parseConfig()`.

File: tests/contracts/basetests.ts (hunk 1 above)
Comment (geofmureithi) [reply tests/contracts/basetests.ts]: Actually, we may want to introspect when something fails, so we might prefer to use `Promises.allSettled`

**Review by eloylp — COMMENTED** [review]
Hello ! thanks for PR description. I am still reading the code in detail, just providing early feedback here:

* Would it be good to add a `README.md` inside the `test` folder, indicating what does this tool and some examples ? Imagine someone with no context lands in that folder, what would be useful ? Also, should we point to this README.md from the root one ?

* Its sad we cannot implement this in Rust to be honest, taking into account we are targeting Soroban smart contracts. I would like to know how far we would be from this and what we would need from Soroban to implement/fix. Maybe its something doable to open an upstream issue :)

**Comment by mariopil** [comment]
Should with close this PR without merging now when the `cargo make` tool will be used for integration tests?

**Comment by geofmureithi** [comment]
Lets close this one.

**Comment by mariopil** [comment]
There is another automated tests tool written using `cargo make `. This one is not needed anymore.
